lib/ledger.js

```diff
@@ -358,6 +358,141 @@ class LedgerBTCApp {
 
     return APDUResponse.hashSign(res).data;
   }
+
+  /**
+   * Prepare message for signing (legacy).
+   * @param {Number[]} path
+   * @param {Buffer} message
+   */
+
+  async signMessagePrepare0(path, message) {
+    assert(this.device);
+    assert(Array.isArray(path), 'Path must be string or array');
+    assert(Buffer.isBuffer(message), 'Message is not a buffer.');
+
+    const data = bufio.write();
+    data.writeBytes(util.encodePath(path));
+    data.writeU8(message.length);
+    data.writeBytes(message);
+
+    assert(data.getSize() <= 255, 'Message + path is too big.');
+
+    const packet = APDUCommand.signMessage(data.render(), true, true, true);
+    const rawResponse = await this.device.exchange(packet.toRaw());
+    const response = APDUResponse.signMessage0(rawResponse);
+
+    return response.data;
+  }
+
+  /**
+   * Prepare message for signing.
+   * @param {Number[]} path
+   * @param {Buffer} message
+   */
+
+  async signMessagePrepare(path, message) {
+    assert(this.device);
+    assert(Array.isArray(path), 'Path must be string or array');
+    assert(Buffer.isBuffer(message), 'Message is not a buffer.');
+
+    let buffers;
+
+    {
+      const data = bufio.write();
+
+      data.writeBytes(util.encodePath(path));
+      data.writeU16BE(message.length);
+      data.writeBytes(message);
+
+      buffers = utilTX.splitBuffer(
+        data.render(),
+        255
+      );
+    }
+
+    const packets = [
+      APDUCommand.signMessage(buffers.shift(), true, true)
+    ];
+
+    for (const buffer of buffers)
+      packets.push(APDUCommand.signMessage(buffer, true, false));
+
+    const last = packets.pop();
+
+    let buffer = Buffer.alloc(0);
+
+    for (const packet of packets) {
+      const data = await this.device.exchange(packet.toRaw());
+      const response = APDUResponse.signMessage(data, buffer);
+      buffer = response.data.encryptedOutputData;
+    }
+
+    const data = await this.device.exchange(last.toRaw());
+    const response = APDUResponse.signMessage(data, buffer, true);
+
+    return response.data;
+  }
+
+  /**
+   * Get the signature.
+   * @param {String} pin (not tested.)
+   * @returns {Buffer}
+   */
+
+  async signMessageFinal(pin) {
+    let data;
+
+    if (pin) {
+      const bw = bufio.write();
+      const buf = Buffer.from(pin, 'utf8');
+
+      bw.writeU8(buf.length);
+      bw.writeBytes(buf);
+
+      data = bw.render();
+    } else {
+      data = Buffer.alloc(1, 0x00);
+    }
+
+    const packet = APDUCommand.signMessage(data, false, false);
+    const rawResponse = await this.device.exchange(packet.toRaw());
+    const response = APDUResponse.signMessageFinal(rawResponse);
+
+    return response.data;
```

Comment:
ledger.js returns `v`, `r`, and `s` instead of just the signature

Reply:
yeah, I was working with python API that just returns buffer, I guess it makes sense to match ledgerjs?

Reply:
I think a common usecase will be ecrecover + verify, when r and s are returned I do:

```
const sig = await bcoinApp.signMessage(message, path);
const rs = Buffer.concat([sig.r,sig.s]);
const param = sig.v;
const pubkey = secp256k1.recover(message, rs, param);
const valid = secp256k1.verify(message, rs, pubkey);
```

Reply:
Providing `verifyMessage` will be more convenient for users.